### PR TITLE
Suggested actions sample

### PR DESCRIPTION
**Pull request by JasonSowers**

### CODE REVIEW

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/appsettings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+﻿{  
+  "MicrosoftAppId": "",
+  "MicrosoftAppPassword": ""
```
Comment (vishwacsena) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/appsettings.json]: These should be moved to the .bot file. The only content in appsettings should be botFilePath and botFileSecret.

File: csharp_dotnetcore/8.Suggested-Actions/BotBuilder.ruleset [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/BotBuilder.ruleset]: I see this file twice. Once in "8.Suggested-Actions" and once in "8.Suggested_Actions". 

Please remove the incorrect file .

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                .SetBasePath(env.ContentRootPath)
+                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
+                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
+                .AddEnvironmentVariables();
+
+            this.Configuration = builder.Build();
+        }
+
+        /// <summary>
+        /// Gets the configuration that represents a set of key/value application configuration properties.
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Please add a "see" tag for IConfiguration here.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        /// </summary>
+        /// <value>
+        /// The IConfiguration that represents a set of key/value application configuration properties.
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Add a See tag here.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        /// </value>
+        public IConfiguration Configuration { get; }
+
+        /// <summary>
+        /// This method gets called by the runtime. Use this method to add services to the container.
+        /// </summary>
+        /// <param name="services">Specifies the contract for a collection of service descriptors.</param>
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Add a See tag here for the services -> IServiceCollection

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddBot<SuggestedActionsBot>(options =>
+            {
+                options.CredentialProvider = new ConfigurationCredentialProvider(this.Configuration);
+
+                // The Memory Storage used here is for local bot debugging only. When the bot
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Consider tightning this language. Perhaps:
"MemoryStorage is for local bot debugging only. When the bot [...]

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                // is restarted, anything stored in memory will be gone.
+                IStorage dataStore = new MemoryStorage();
+
+                // For production bots use the Azure Blob or
+                // Azure CosmosDB storage provides, as seen below. To include any of
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: sp. provides => providers.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                // the Azure based storage providers, add the Microsoft.Bot.Builder.Azure
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Change "[...] any of the Azure [...]" just "To include Azure based"

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                // Nuget package to your solution. That package is found at:
+                // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // Uncomment this line to use blob storage
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Azure Blog Storage

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                // Create and add conversation state.
+                var convoState = new ConversationState(dataStore);
+                options.State.Add(convoState);
+
+                // Add state to BotStateSet middleware (that require auto-save).
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: I'm not clear on what the comment is trying to tell me.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                // Note: Developers may choose not to add all the state providers to this middleware if save is not required.
+                var stateSet = new BotStateSet(options.State.ToArray());
+                options.Middleware.Add(stateSet);
+            });
+        }
+
+        /// <summary>
+        /// This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        /// </summary>
+        /// <param name="app">The application builder.  This provides the mechanisms to configure an application's request pipeline.</param>
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Add See tags here.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+
+namespace Suggested_Actions
+{
+    /// <summary>
+    /// This bot will respond to the user's input suggested actions.
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: This comment means something *ONLY* if I already know what input suggested actions are. In this case, I think we should add more here.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    /// </summary>
+    public class SuggestedActionsBot : IBot
+    {
+        /// <summary>
+        /// This controls what happens when an activity gets sent to the bot.
+        /// </summary>
+        /// <param name="turnContext">Provides the context for the turn of the bot.</param>
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Add a see tag for ITurnContext

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: see tag

Comment (cleemullins) [reply csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: also: "(Optional)":

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            if (turnContext == null)
+            {
+                throw new ArgumentNullException(nameof(turnContext));
+            }
+
+            switch (turnContext.Activity.Type)
+            {
+                case ActivityTypes.Message:
+                    var text = turnContext.Activity.Text.ToLowerInvariant();
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Why ToLowerInvariant()? A comment explaining why this was chosen would be good.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/appsettings.json (hunk 1 above)
Comment (JasonSowers) [reply csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/appsettings.json]: Roger that

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+                    switch (text)
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: I'm not a huge fan of these nested switch statements. Consider making this a method to be more clear.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+                                $"Please select a color from the suggested actions",
+                                cancellationToken: cancellationToken);
+                            break;
+                        }
+                    }
+
+                    await SendSuggestedActionsAsync(turnContext, cancellationToken);
+                    break;
+                case ActivityTypes.ConversationUpdate:
+
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Extra blank line

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+                    // Send a welcome message to the user and tell them what actions they need to perform to use this bot
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Good comment. Needs a "." at the end. 
Also, "need to perform" -> "may perform"

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+                    if (turnContext.Activity.MembersAdded.Any())
+                    {
+                        {
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Extra {. 
Extract to method.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+                    }
+
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Creates and sends an activity with suggested actions to the user.
+        /// </summary>
+        /// <param name="turnContext">Provides the context for the turn of the bot.</param>
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: see tags for ITurnContext

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: (Optional), add see tag as well.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> representing the operation result of the operation.</returns>
+        private static async Task SendSuggestedActionsAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            var reply = turnContext.Activity.CreateReply("What is your favorite color");
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Add a "?" to the end of the text sent to the user.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+            reply.SuggestedActions = new SuggestedActions()
+            {
+                Actions = new List<CardAction>()
+                {
+                    new CardAction() { Title = "Red", Type = ActionTypes.ImBack, Value = "Red" },
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Comments. For example, what is ImBack and why was it chosen?

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Suggested_Actions.csproj [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+  <ItemGroup>
+    <PackageReference Include="AdaptiveCards" Version="1.0.3" />
+    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.5" />
+    <PackageReference Include="Microsoft.Bot.Builder" Version="4.0.0.38511" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.0.0.38511" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.0.0.38511" />
+    <PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003" PrivateAssets="all" />
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta008" PrivateAssets="all" />
+  </ItemGroup>
+
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Suggested_Actions.csproj]: Extra blank line

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/appsettings.json [hunk 25]
```diff
@@ -0,0 +1,4 @@
+﻿{  
+  "MicrosoftAppId": "",
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/appsettings.json]: These are in the Bot File now. I believe we should remove this file.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs (hunk 23 above)
Comment (JasonSowers) [reply csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Added these comments to the summary section of this method.  It made more sense there as I could link ActionTypes via a 'see' reference

**Comment by JasonSowers** [comment]
@cleemullins can you look at [my latest commit](https://github.com/Microsoft/BotBuilder-Samples/pull/303/commits/f1b9b0526dd4e009ae05b5a907af48535ed8101d), there seems to be some sort of disconnect.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs [hunk 26]
```diff
@@ -35,39 +35,38 @@ public Startup(IHostingEnvironment env)
...
                 // is restarted, anything stored in memory will be gone.
                 IStorage dataStore = new MemoryStorage();
 
                 // For production bots use the Azure Blob or
-                // Azure CosmosDB storage provides, as seen below. To include any of
-                // the Azure based storage providers, add the Microsoft.Bot.Builder.Azure
+                // Azure CosmosDB storage providers, as seen below. To the Azure
+                // based storage providers, add the Microsoft.Bot.Builder.Azure
                 // Nuget package to your solution. That package is found at:
                 // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/Startup.cs]: Please make this URL use the standard XML Comment syntax for URLs.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 27]
```diff
@@ -12,17 +12,27 @@
...
+    /// Suggested actions enable your bot to present buttons that the user
+    /// can tap to provide input. Suggested actions appear close to the composer
+    /// and enhance user experience by enabling the user to answer a question or
+    /// make a selection with a simple tap of a button, rather than having to
+    /// type a response with a keyboard. Unlike buttons that appear within rich
+    /// cards (which remain visible and accessible to the user even after being tapped),
+    /// buttons that appear within the suggested actions pane will disappear after
+    /// the user makes a selection. This prevents the user from tapping stale buttons
+    /// within a conversation and simplifies bot development (since you will not need
+    /// to account for that scenario).
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: This is *awesome*. 👍

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 28]
```diff
@@ -33,79 +43,100 @@ public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancel
             switch (turnContext.Activity.Type)
             {
                 case ActivityTypes.Message:
+                    // Extract the text from the message activity the user sent
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: "." at the end.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 29]
```diff
@@ -33,79 +43,100 @@ public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancel
...
+                    // Make this lowercase not accounting for culture in this case
+                    // so that there are fewer string variations which you will have
+                    // have to account for in your bot.
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Consider cleaning up the text. The grammar is awkward.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 30]
```diff
@@ -33,79 +43,100 @@ public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancel
...
 
-                        case "blue":
-                        {
-                            await turnContext.SendActivityAsync(
-                                "Blue is the best color, I agree.",
-                                cancellationToken: cancellationToken);
-                            break;
-                        }
+                    // Take the input from the user and create the appropriate response.
+                    var responseText = ProcessInput(text);
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Nice.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 31]
```diff
@@ -33,79 +43,100 @@ public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancel
...
-                                    await turnContext.SendActivityAsync(
-                                        $"Welcome to SuggestedActionsBot {member.Name}." +
-                                        $" This bot will introduce you to suggestedActions." +
-                                        $"  Please answer the question:",
-                                        cancellationToken: cancellationToken);
-                                    await SendSuggestedActionsAsync(turnContext, cancellationToken);
-                                }
-                            }
-                        }
+                        await SendWelcomeMessageAsync(turnContext, cancellationToken);
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Nice.

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs [hunk 32]
```diff
@@ -33,79 +43,100 @@ public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancel
...
                     }
 
                     break;
             }
         }
 
         /// <summary>
-        /// Creates and sends an activity with suggested actions to the user.
+        /// On a conversation update activity sent to the bot, the bot will
```
Comment (cleemullins) [inline csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Consider something like this: "When a conversation update activity is sent to the bot, the bot should send a welcome messages to any new user(s)."

**Comment by cleemullins** [comment]
Sure. It’ll be a few hours though. Kids / bed time / dinner / school tomorrow....


________________________________

File: csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs (hunk 27 above)
Comment (JasonSowers) [reply csharp_dotnetcore/8.Suggested-Actions/Suggested_Actions/SuggestedActionsBot.cs]: Thank you

File: csharp_dotnetcore/6.Using-Cards/Using_Cards.sln [hunk 33]
```diff
@@ -0,0 +1,25 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (JasonSowers) [inline csharp_dotnetcore/6.Using-Cards/Using_Cards.sln]: This file 👎

**Comment by cleemullins** [comment]
Let's baseline this. There may be some additional comments and changes moving forward, but this is an awesome starting point!!

**Pull request merged**